libazureinit/src/provision/mod.rs

```diff
@@ -0,0 +1,342 @@
+            hostname_backends: self.hostname_backends,
+            user_backends: self.user_backends,
+            password_backends: self.password_backends,
+            state: PhantomData,
+        }
+    }
+}
+
+impl Provision<PasswordBackend> {
+    pub fn password_provisioners(
+        mut self,
+        backend: impl Into<Vec<password::Provisioner>>,
+    ) -> Provision<AllBackends> {
+        self.password_backends = backend.into();
+        Provision {
+            hostname: self.hostname,
+            username: self.username,
+            keys: self.keys,
+            password: self.password,
+            hostname_backends: self.hostname_backends,
+            user_backends: self.user_backends,
+            password_backends: self.password_backends,
+            state: PhantomData,
+        }
+    }
+}
+
+impl Provision<AllBackends> {
+    /// Set a password for the user being created.
+    ///
+    /// If a password is set, the caller must also set at least one provisioner with
+    /// [`Provision<PasswordBackend>::password_provisioners`].
+    pub fn password(mut self, password: String) -> Provision<PasswordBackend> {
+        self.password = Some(password);
+        Provision {
+            hostname: self.hostname,
+            username: self.username,
+            keys: self.keys,
+            password: self.password,
+            hostname_backends: self.hostname_backends,
+            user_backends: self.user_backends,
+            password_backends: self.password_backends,
+            state: PhantomData,
+        }
+    }
+
+    /// Add the provided SSH keys to the authorized key file of the user being provisioned.
+    pub fn ssh_keys(mut self, keys: impl Into<Vec<PublicKeys>>) -> Self {
+        self.keys = keys.into();
+        self
+    }
+
+    /// Apply the selected configuration to the host.
+    #[instrument(skip_all)]
+    pub fn provision(self) -> Result<(), Error> {
+        self.user_backends
+            .into_iter()
+            .find_map(|backend| backend.create(&self.username).map_err(|e| {
+                    tracing::warn!(error=?e, backend=?backend, resource="user", "Provisioning failed");
+                    e
+            }).ok())
+            .ok_or(Error::NoUserProvisioner)?;
+
+        self.password_backends
+            .into_iter()
+            .find_map(|backend| {
+                backend
+                    .set(&self.username, self.password.as_deref().unwrap_or(""))
+                    .map_err(|e| {
+                        tracing::warn!(error=?e, backend=?backend, resource="password", "Provisioning failed");
+                        e
+                    })
+                    .ok()
+            })
+            .ok_or(Error::NoPasswordProvisioner)?;
+
+        if !self.keys.is_empty() {
+            let user = nix::unistd::User::from_name(&self.username)?
+                .ok_or_else(|| Error::UserMissing {
+                    user: self.username,
+                })?;
```

Comment:
In case of building with Rust 1.70.0, clippy prints out the warning:

```
warning: unnecessary closure used to substitute value for `Option::None`
   --> libazureinit/src/provision/mod.rs:160:24
    |
160 |               let user = nix::unistd::User::from_name(&self.username)?
    |  ________________________^
161 | |                 .ok_or_else(|| Error::UserMissing {
162 | |                     user: self.username,
163 | |                 })?;
    | |__________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
    = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default
help: use `ok_or(..)` instead
    |
161 ~                 .ok_or(Error::UserMissing {
162 +                     user: self.username,
163 ~                 })?;
    |
```